README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# shoredate: An R package for shoreline dating coastal Stone Age sites

<!-- badges: start -->
<!-- [![CRAN Version](http://www.r-pkg.org/badges/version/shoredate)](https://cran.r-project.org/package=shoredate) -->
<!-- [![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0) -->
<!-- [![R-CMD-check](https://github.com/isakro/shoredate/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/isakro/shoredate/actions/workflows/R-CMD-check.yaml) -->
<!-- [![Codecov test coverage](https://codecov.io/gh/isakro/shoredate/branch/master/graph/badge.svg)](https://app.codecov.io/gh/isakro/shoredate?branch=master) -->
<!-- [![DOI](https://joss.theoj.org/papers/10.21105/joss.05337/status.svg)](https://doi.org/10.21105/joss.05337) -->
<!-- badges: end -->

The package *shoredate* offers methods to shoreline date coastal Stone
Age sites based on their present-day elevation and the trajectory of
past relative sea-level change. Shoreline dating is based on the premise
that coastal Stone Age sites in large parts of Fennoscandia were located
on or close to the shoreline when they were in use. The package and
method as implemented here was originally developed for the Norwegian
Skagerrak coast in south-eastern Norway, based on an empirically derived
estimate of the likely elevation of sites above sea-level when they were
occupied (Roalkvam 2023).

While the package offers ways to extend and adjust the method for
application in other areas, the ways and degree to which the procedures
are directly applicable elsewhere is largely undetermined and likely to
vary between contexts. Furthermore, such extensions are currently
limited to regions that have been characterised by a monotonic
trajectory of relative sea-level regression. Do also note that as the
method is dependent on regularities in human behaviour, and as the
Roalkvam (2023) study provides an initial formalisation of the method,
it is hefted with unexplored uncertainties, also within the core area
for which it was developed. In sum therefore, the dates achieved with
the package should be treated with care.

## Installation and loading

*shoredate* can be installed from
[CRAN](https://cran.r-project.org/package=shoredate) with:

``` r
install.packages("shoredate")
```

The latest development version can be installed from
[GitHub](https://github.com/isakro/shoredate) using `devtools`:

``` r
# install.packages("devtools")
devtools::install_github("isakro/shoredate", build_vignettes = TRUE)
```

When it has been installed the package can be loaded:

``` r
library(shoredate)
```

## Geographical and temporal coverage

As the method of shoreline dating is determined by relative sea-level
change, it is dependent on reliable geological reconstructions of this
development. The method as outlined here was originally developed for
the Skagerrak region of south-eastern Norway – extending from Horten in
the north east to Arendal in the south west (see figure below). This
region has newly compiled shoreline displacement curves for Horten
(Romundset 2021), Porsgrunn (Sørensen et al. 2014; Sørensen et
al. 2023), Tvedestrand (Romundset 2018; Romundset et al. 2018) and
Arendal (Romundset 2018). The region also formed the study area for
Roalkvam (2023), in which the method and its parameters were derived.
The remainder of this document and the [main
vignette](https://isakro.github.io/shoredate/articles/shoredate.html)
focuses on this area. It is, however, possible to adapt the procedures
for application in other regions, which is outlined in the [second
vignette](https://isakro.github.io/shoredate/articles/extending-shoredate.html).

The first figure below shows the location of the spatial limit within
which the method was originally derived. The shoreline isobases in the
second figure represent contours along which the shoreline displacement
has followed the same trajectory. These correspond to the displacement
curves and place names in the third figure, which also indicates the
temporal coverage of the method within the region.

<img src="man/figures/coverage.png" width="90%" height="90%" style="display: block; margin: auto;" />

As human occupation only occurred some time after the retreat of the
Fennoscandian Ice Sheet, the currently oldest known sites in Norway are
from around 9300 BCE (e.g. Glørstad 2016). Although no sites are yet
known to be that old, the earliest possible age to achieve with
*shoredate* in the Skagerrak region of south-eastern Norway is 9469 BCE,
which marks the latest start date among the employed displacement
curves. If a site has an elevation that implies a date older than this,
it is returned as NA and a warning is given.

In Roalkvam (2023) it was found that sites in the study region tend to
be located on or close to the shoreline up until around the transition
to the Late Neolithic, c. 2500 BCE, which thus marks the upper limit for
the applicability of the method in the region. A date that has a later
start date than this is therefore returned as NA with a warning, with
the default settings. The temporal range is indicated by the dashed
lines in the plot above that displays the shoreline displacement curves.
Additionally, if the probability of a date extends beyond 1950 CE (0 cal
BP), thus indicating a site location below the present-day sea-level,
this overshooting probability is cut off and the date is normalised to
sum to unity.

## Interpolating shoreline displacement to a site location

To shoreline date a site, a reconstruction of local shoreline
displacement is necessary. There are currently four reliable geological
displacement curves available from within the study area. Each of these
is associated with a shoreline isobase, along which the trajectory of
relative sea-level change has been the same. To find the local
displacement curve, the curves are interpolated to a site location using
inverse distance weighting, where the default is to weigh the
interpolation by the square of the inverse distance between site and
isobases.

``` r
# Create example point using the coordinate system required for the 
# interpolation procedure: WGS84 / UTM zone 32N (EPSG: 32632)
target_point <- sf::st_sfc(sf::st_point(c(538310, 6544255)), crs = 32632)

# Create a simple map showing the target location relative to the isobases
# of the displacement curves
target_plot(target_point)
```

<img src="man/figures/target_location-1.png" style="display: block; margin: auto;" />

``` r
# Interpolate shoreline displacement curve for the target location
target_curve <- interpolate_curve(target_point)

# Plot displaying the interpolated curve, reducing the opacity of the
# geological displacement curves.
displacement_plot(target_curve, displacement_alpha = 0.4)
```

<img src="man/figures/interpolated_curve-1.png" style="display: block; margin: auto;" />

Unless a pre-compiled curve is provided, this interpolation is performed
under the hood for each site when calling `shoreline_date()`, which is
the function that performs the shoreline dating procedure.

## Shoreline dating a site

Below is a basic example outlining how to date a single site by manually
specifying the site elevation. The default settings are used for the
dating procedure and for plotting the resulting shoreline date.

``` r
# Using the example point from above and specifying it's elevation
target_date <- shoreline_date(site = target_point, elevation = 70)

# Call to plot
shoredate_plot(target_date)
```

<img src="man/figures/date-1.png" style="display: block; margin: auto;" />

The blue gamma distribution on the y-axis represents the likely
elevation of the site above sea-level when it was in use, which is
described by an empirically derived gamma function with the parameters
$\alpha$ (shape) = 0.286 and $\sigma$ (scale) = 20.833. This starts from
the elevation of the site. The red envelope is the shoreline
displacement curve as interpolated to the site location. The probability
from the gamma distribution is transferred to the calendar scale using
the displacement curve. This gives the resulting shoreline date in grey,
which is underlined by the 95% highest density region (HDR) in black
(see Roalkvam 2023 for more details). By default, the shoreline date is
normalised to sum to unity. The default resolution on the calendar scale
is 10 years.

Calling the date object, which has the custom class `shoreline_date`,
prints the name of the site, its elevation and the HDR:

``` r
target_date
#> ===============
#> Site:  1
#> Elevation:  70 
#> 
#> 95% HDR:
#> 8610 BCE-6910 BCE
#> 5880 BCE-5100 BCE
```

The first column of a data frame beyond the geometry of the spatial
objects will be taken to represent site names. If no such column exist,
the sites are simply numbered as they are passed to `shoreline_date()`.

## Further documentation

The procedures outlined above have focused on the basic functions and
default behaviours of the package when dating a single site in the area
for which the package was originally developed. For further usage and a
more detailed walk-through, see the main vignette by calling
`vignette("shoredate")` or by accessing it on the
[website](https://isakro.github.io/shoredate/articles/shoredate.html).

Furthermore, a second vignette which can be accessed with
`vignette("extending-shoredate")` or viewed on the
[website](https://isakro.github.io/shoredate/articles/extending-shoredate.html),
builds on the main vignette and outlines ways in which the package can
be applied to other regions.

# References

Glørstad, H. 2016. Deglaciation, sea-level change and the Holocene
colonization of Norway. *Geological Society, London, Special
Publications* 411(1):9–25. DOI:
[10.1144/SP411.7](https://doi.org/10.1144/SP411.7)

Romundset, A. 2018. Postglacial shoreline displacement in the
Tvedestrand–Arendal area. In *The Stone Age Coastal Settlement in
Aust-Agder, Southeast Norway*, edited by Reitan, G. and Sundström, L.
Cappelen Damm Akademisk, Oslo, pp. 463–478. DOI:
[10.23865/noasp.50](https://doi.org/10.23865/noasp.50)

Romundset, A. 2021. *Resultater fra NGUs undersøkelse av etteristidas
strandforskyvning nord i Vestfold*. Geological Survey of Norway,
Trondheim.

Romundset, A., Lakeman, T.R. and Høgaas, F. 2018. Quantifying variable
rates of postglacial relative sea level fall from a cluster of 24
isolation basins in southern Norway. *Quaternary Science Reviews*
197:175e192. DOI: [10.1016/
j.quascirev.2018.07.041](https://www.sciencedirect.com/science/article/pii/S0277379118302890)

Roalkvam, I. 2023. A simulation-based assessment of the relation between
Stone Age sites and relative sea-level change along the Norwegian
Skagerrak coast. *Quaternary Science Reviews* 299:107880. DOI:
[10.1016/j.quascirev.2022.107880](https://doi.org/10.1016/j.quascirev.2022.107880)

Sørensen, R, Henningsmoen, K.E. Høeg, H.I. and Gälman V. 2023. Holocen
vegetasjonshistorie og landhevning i søndre Vestfold og sørøstre
Telemark. In *The Stone Age in Telemark. Archaeological Results and
Scientific Analysis from Vestfoldbaneprosjektet and E18
Rugtvedt–Dørdal*, edited by Persson, P. and Solheim, S., in press.

Sørensen, R, Henningsmoen, K.E. Høeg, H.I. and Gälman V. 2014. Holocene
landhevningsstudier i søndre Vestfold og sørøstre Telemark – Revidert
kurve. In *Vestfoldbaneprosjektet. Arkeologiske undersøkelser i
forbindelse med ny jernbane mellom Larvik og Porsgrunn. Bind 1*, edited
by Melvold, S. and Persson, P. Portal forlag, Kristiansand, pp. 36–47.
DOI: [10.23865/noasp.61](https://doi.org/10.23865/noasp.61)

# Contributing

Contributions and suggestions for improvement are all very welcome.
Instructions for contributing can be found in the [Guide to
Contributing](https://github.com/isakro/shoredate/blob/master/CONTRIBUTING.md).
Please note that this project is released with a [Contributor Code of
Conduct](https://github.com/isakro/shoredate/blob/master/.github/CODE_OF_CONDUCT.md).
By participating in this project you agree to abide by its terms.


NEWS.md
# shoredate (development version)

# shoredate 1.1.1 (2023-10-23)
- Added the parameter `date_probability_scale` to `shoredate_plot()` to make
 it possible to adjust the scaling of the probability distribution of the 
 shoreline date to better fit the plot.
 - Removed dependency on `rgdal` (@rsbivand, #9)

## shoredate 1.1.0 (2023-05-22)
- Added a vignette that demonstrates how the package can be applied to other
 regions outside south-eastern Norway (@benmarwick, #6).
- Removed the function `check_target_location()` and set up more specific and
 appropriate checks within `interpolate_curve()` and `target_plot()`.
- Added the package `rnaturalearth` to suggests in DESCRIPTION. This can be used
 with `target_plot()` to create a map for any area of the world, extending the 
 transferability of the package (@benmarwick, #6).
- Made most graphical parameters in `shoredate_plot()`, `displacement_plot()` 
 and `target_plot()` variables that be adjusted by the user (@kanishkan91, #5).
- Adjusted how the `geom` column of `sf` objects passed to `target_plot()` are 
 evaluated (@kanishkan91, #2).
- Cleaned up code to reduce the amount of `@Import` and `@ImportFrom` in 
 NAMESPACE, opting instead for the `package::function()` syntax
 (@kanishkan91, #4).
- Set up a pkgdown site for the package, served through GitHub at 
  https://isakro.github.io/shoredate/ (@kanishkan91, #3).

## shoredate 1.0.2 (2023-05-02)
 - `shoredate_hdr()` and `shoreline_date()` now returns the weighted mean date.
 - `shoreline_date()` now has the parameter model, which accepts either "none" 
 or "gamma". If "none" is specified then the distance between site and shoreline
 is assumed to be zero.
 - Now using the package `ggspatial` for scale bar in maps rather than `ggsn`, 
 due  to the retiring of r-spatial packages (@rsbivand, #1).

## shoredate 1.0.1 (2023-03-06)

 - `sum_shoredates()` can now handle NA dates.
 - `shoreline_date()` now accepts a vector of site names for the argument 
 `sites` if interpolated displacement curves are also provided.

## shoredate 1.0.0 (2023-01-19)
First release.


R/shoreline_date.R
#' Shoreline date
#'
#' A function for shoreline dating Stone Age sites based on their present-day
#'  elevation, their likely elevation above sea-level when in use and the
#'  trajectory of past shoreline displacement. Details and caveats pertaining to
#'  the implemented method is given in Roalkvam (2023).
#'
#' @param sites Vector giving one or more site names, or, if displacement curves
#'   are to be interpolated, objects of class `sf` representing the sites to be
#'   dated. In the case of a spatial geometry, the first column is taken as the
#'   site name.
#' @param elevation Vector of numeric elevation values for each site or a
#'  an elevation raster of class `SpatRaster` from the package
#'  `terra` from where the elevation values are to be derived.
#' @param elev_reso Numeric value specifying the resolution with which to step
#'  through the distribution representing the distance between site and
#'  shoreline. Defaults to 0.01m.
#' @param cal_reso Numeric value specifying the resolution to use on the
#'   calendar scale. Defaults to 10.
#' @param isobase_direction A vector of numeric values defining the direction(s)
#'   of the isobases. Defaults to 327.
#' @param sum_isobase_directions Logical value indicating that if multiple
#'  isobase directions are specified in `isobase_direction` the results should
#'  be summed for each site using `sum_shoredates`. Defaults to FALSE.
#' @param model Character vector specifying the statistical model with which to
#'  model the distance from site to shoreline. Currently accepts either "none"
#'  or "gamma". Defaults to "gamma".
#' @param model_parameters Vector of numeric values specifying the parameters
#'  for the statistical model describing the distance between site and
#'  shoreline. Defaults to c(0.286, 20.833), denoting the shape and scale of the
#'   default gamma function, respectively.
#' @param elev_fun Statistic to define site elevation if this is to be derived
#'  from an elevation raster. Uses `terra::extract()`. Defaults to mean.
#' @param upper_temp_limit Numerical value giving the upper temporal limit.
#'  Dates with a start date after the limit are returned as NA. Defaults to
#'  -2500, i.e. 2500 BCE.
#' @param target_curve Data frame holding pre-computed shoreline displacement
#'  curve. This has to have the same or higher resolution on the calendar scale
#'  as that specified with `cal_reso`. [interpolate_curve()] will be run if
#'  nothing is provided to `target_curve`. Defaults to NA.
#' @param hdr_prob Numeric value specifying the coverage of the highest density
#'  region. Defaults to 0.95.
#' @param normalise Logical value specifying whether the shoreline date should
#'  be normalised to sum to unity. Defaults to TRUE.
#' @param sparse Logical value specifying if only site name and shoreline date
#'  should be returned. Defaults to FALSE. Note that of the functions for
#'  further treatment, sparse dates are only compatible with
#'  [sum_shoredates()].
#' @param verbose Logical value indicating whether progress should be printed to
#'  console. Defaults to FALSE.
#'
#' @return A nested list of class `shoreline_date` holding the shoreline date
#'  results and associated metadata for each dated site for each isobase
#'  direction. The elements of each date is:
#'
#'  * `site_name` name of the site.
#'  * `site_elev` elevation of the site.
#'  * `date` data frame with the columns `bce` where negative values
#'  indicate years BCE and positive CE, as well as `probability`, which gives
#'  the probability mass for each year.
#'  * `weighted_mean` the weighted mean date.
#'  * `hdr_start` start values for the HDR ranges.
#'  * `hdr_end` end values for the HDR ranges.
#'  * `hdr_prob` probability level for the HDR.
#'  * `dispcurve` data frame holding the displacement curve used for dating
#'  the site. This has the columns `bce`, giving years BCE/CE. `lowerelev`,
#'  the lower limit for the elevation of the shoreline for each year.
#'  `upperelev`, the upper limit for elevation of the shoreline for each year.
#'  * `dispcurve_direction` direction of the isobases in use.
#'  * `model_parameters` parameters for the statistical model.
#'  * `modeldat` data frame holding the model distribution. The column
#'  `offset` denotes the vertical distance (m) from the shoreline, as specified
#'  by the `elev_reso` argument. `px` is the cumulative probability at each step
#'  of `offset`, and `probs` is the probability of each step found by
#'  subtracting the preceding value from each value of `px`.
#'  * `cal_reso` resolution on the calendar scale.
#'
#' @export
#'
#' @references
#' Roalkvam, I. 2023. A simulation-based assessment of the relation between
#'  Stone Age sites and relative sea-level change along the Norwegian Skagerrak
#'  coast. \emph{Quaternary Science Reviews} 299:107880. DOI:
#'  https://doi.org/10.1016/j.quascirev.2022.107880
#'
#' @examples
#' # Create example point using the required CRS WGS84 UTM32N (EPSG: 32632)
#' target_point <- sf::st_sfc(sf::st_point(c(538310, 6544255)), crs = 32632)
#'
#' # Date target point, manually specifying the elevation instead of providing
#' # an elevation raster. Reducing elev_reso and cal_reso for speed.
#' shoreline_date(sites = target_point,
#'                elevation = 80,
#'                elev_reso = 1,
#'                cal_reso = 400)
shoreline_date <- function(sites,
                           elevation = NA,
                           elev_reso = 0.01,
                           cal_reso = 10,
                           isobase_direction = 327,
                           sum_isobase_directions = FALSE,
                           model = "gamma",
                           model_parameters = c(0.286, 20.833),
                           elev_fun = "mean",
                           upper_temp_limit = -2500,
                           target_curve = NA,
                           hdr_prob = 0.95,
                           normalise = TRUE,
                           sparse = FALSE,
                           verbose = FALSE){

  # Make sfc geometries be represented as a sf data frame
  if(!inherits(sites, c("sf", "data.frame"))){
    if (all(is.na(target_curve)) & inherits(sites, "sfc")) {
      sites <- sf::st_as_sf(sites, crs = sf::st_crs(sites))
      # Make vector of site names a data frame for nrow() below
    } else {
      sites <- as.data.frame(sites)
    }
  }

  if (!(is.numeric(elevation) |
        (inherits(elevation, c("raster","SpatRaster"))))){
    stop("Numeric values specifying the site elevations or an elevation raster must be provided.")
  }

  if(is.numeric(elevation) & length(elevation) != nrow(sites)){
    stop(paste("Specify one elevation value per site.", length(elevation),
               "elevation values and", nrow(sites), "sites were provided."))
  }

  # If isobase directions other than the default is provided, create these
  if (all(is.na(target_curve)) & any(isobase_direction != 327)) {
    isobases <- create_isobases(isobase_direction)
    # If the default is used, load the precompiled isobases
  } else{
    isobases <- sf::st_read(
      system.file("extdata/isobases.gpkg",
                  package = "shoredate",
                  mustWork = TRUE), quiet = TRUE)
  }

  # List to hold dates
  shorelinedates <- list()
  for (i in 1:nrow(sites)) {

    if (verbose) {
      print(paste("Site", i, "of", nrow(sites)))
    }

    # If a target curve is not provided and multiple isobase directions have
    # been specified, interpolate these
    if (all(is.na(target_curve)) &
        length(unique(isobases$direction)) > 1) {

      sitecurve <- interpolate_curve(target = sites[i,],
                                     isobases = isobases, cal_reso = cal_reso,
                                     verbose = verbose)
      # If default isobase direction is to be used, but no interpolated curve
      # is provided
    } else if (all(is.na(target_curve))) {
      sitecurve <- interpolate_curve(target = sites[i,],
                                     isobases = isobases, cal_reso = cal_reso,
                                     verbose = verbose)
    } else {
      # If pre-compiled/interpolated curve(s) are provided
      sitecurve <- target_curve
    }

    if (inherits(sitecurve, "list")) {
      sitecurve <- do.call(rbind.data.frame, sitecurve)
    }

    # If there is no direction to the provided target curve, specify that this
    # is NA.
    if (!("direction" %in% names(sitecurve))) {
      sitecurve$direction <- NA
    }

    # Create sequence of years BCE matching displacement curves
    # (suppressWarsnings() as different lengths should also return FALSE)
    if (suppressWarnings(all((seq(-1950, 10550,
                                  cal_reso) * -1) ==  sitecurve$bce))) {

      bce <- seq(-1950, 10550,  cal_reso) * -1
    } else {
    # A provided target_curve could cover a different time-span, in which
    # case an adjusted bce element is created
      bce <- seq(min(sitecurve$bce), max(sitecurve$bce), cal_reso)
    }

    # Make sure the resolution of the provided curve matches cal_reso.
    # If not, downsample so that it does.
    if ((sitecurve$bce[1]-sitecurve$bce[2]) != cal_reso) {

      upperelev <- stats::approx(sitecurve[,"bce"],
                                 sitecurve[,"upperelev"],
                                 xout = bce)[["y"]]
      lowerelev <- stats::approx(sitecurve[,"bce"],
                          sitecurve[,"lowerelev"],
                          xout = bce)[["y"]]

      sitecurve <- data.frame(bce,
                              upperelev,
                              lowerelev,
                              "direction" = unique(sitecurve$direction))
    }

    if (!inherits(elevation, c("raster", "SpatRaster"))) {
      # if (any(is.na(elevation))) {
      #   return(NA)
      # } else {
      siteelev <- elevation[i]
    } else {
      siteelev <- terra::extract(elevation,
                                 terra::vect(sites[i,]), fun = elev_fun)[,-1]
    }

    # Elevation offsets to step through by increments of elev_reso
    inc <- seq(0, siteelev, elev_reso)

    # Perform one shoreline date per isobase direction.
    date_isobases <- list()
    for (k in 1:length(unique(sitecurve$direction))) {

      if(!(is.na(unique(sitecurve$direction)[k]))){
        temp_curve <- sitecurve[sitecurve$direction ==
                                unique(sitecurve$direction)[k],]
      } else {
        temp_curve <- sitecurve
      }

      # Make sure the displacement curve is sorted descending
      # temp_curve <- temp_curve[order(temp_curve$bce, decreasing = TRUE),]

      # Set up data frame to hold results
      dategrid <- data.frame(
        bce = bce,
        probability = 0)

      # Assign site name
      if (inherits(sites, "sf") &
          ncol(as.data.frame(sites)) == 1) {
        site_name <- as.character(i)
      } else if (inherits(sites, "sf")){
        site_name <- as.character(sf::st_drop_geometry(sites)[i,1])
      } else {
        site_name <- as.character(sites[i, 1])
      }

      # Find oldest possible date
      mdate <- temp_curve[which(temp_curve[,"lowerelev"] ==
                                  max(temp_curve[,"lowerelev"],
                                      na.rm = TRUE)), "bce"]

      # Check that site date is not above this limit
      msdate <- stats::approx(temp_curve[,"lowerelev"],
                              bce, xout = siteelev, method = "constant")[['y']]

      # If it is msdate will be NA and the date is returned as NA with a
      # warning. Do not print isobase direction if this is
      # the default.
      if (is.na(msdate) & is.na(unique(temp_curve$direction))) {
        warning(paste0("The elevation of site ", site_name,
                      " implies an earliest possible date older than ", mdate,
                      " BCE and is out of bounds. The date is returned as NA."))
        dategrid$probability <- NA
        modeldat <- NA
      } else if(is.na(msdate) & unique(temp_curve$direction) == 327) {
        warning(paste0("The elevation of site ", site_name,
                       " implies an earliest possible date older than ", mdate,
                       " BCE and is out of bounds. The date is returned as NA."))
        dategrid$probability <- NA
        modeldat <- NA
      } else if (is.na(msdate)) {
        warning(paste0("The elevation of site ", site_name,
                       " with an isobase direction of ",
                       unique(temp_curve$direction),
                       " implies an earliest possible date older than ", mdate,
                       " BCE and is out of bounds. The date is returned as NA."))
        dategrid$probability <- NA
        modeldat <- NA
      } else {

        if(model == "gamma"){
          # Set up data frame and assign probability to the offset increments
          modeldat <- data.frame(
            offset = inc,
            px = stats::pgamma(inc,
                               shape = model_parameters[1],
                               scale =  model_parameters[2]))
          modeldat$probs <- c(diff(modeldat$px), 0)
          modeldat <- modeldat[modeldat$px < 0.99999,]
        } else if(model == "none"){
          modeldat <- data.frame(
            offset = 0,
            probs = 1
          )
        }

        if (verbose) {
          print("Performing shoreline dating")
          pb <- utils::txtProgressBar(min = 0,
                                      max = nrow(modeldat),
                                      style = 3,
                                      char = "=")
        }

        for (j in 1:nrow(modeldat)) {

          # Subtract offset
          adjusted_elev <- as.numeric(siteelev - modeldat$offset[j])

          # Find lower date
          lowerd <- stats::approx(temp_curve[,"lowerelev"], bce,
                                  xout = adjusted_elev,
                                  method = "constant")[['y']]

          # Find upper date
          upperd <- stats::approx(temp_curve[,"upperelev"], bce,
                                  xout = adjusted_elev,
                                  method = "constant")[['y']]

          # Find youngest and oldest date
          earliest <- min(c(lowerd, upperd))
          latest <- max(c(lowerd, upperd))

          # Add probability to each year in range
          if (!is.na(latest) && !is.na(earliest)) {

            # Identify range, rounding to closest cal_reso
            year_range <- seq(earliest, latest, cal_reso)

            # Identify probability to be distributed across range
            prob <- 1/length(year_range) * modeldat$probs[j]

            # Add probability to calendar years
            dategrid[dategrid$bce %in% year_range, "probability"] <-
              dategrid[dategrid$bce %in% year_range, "probability"] + prob
          }
          if (verbose) {
            utils::setTxtProgressBar(pb, j)
          }
        }
        if (verbose) {
          close(pb)
        }
      }

      # Make the date NA and return a warning if it's latest possible start is
      # younger than upper limit (defaults to 2500 BCE). Also return isobase
      # direction if this is not the default.
      if (!all(is.na(dategrid$probability))) {
        if (min(dategrid$bce[dategrid$probability > 0]) > upper_temp_limit) {
          dategrid$probability <- NA

          if (unique(temp_curve$direction) == 327) {
            warning(paste("Site", site_name,
                          "has a younger possible start date than 2500 BCE and is returned as NA."))
          } else {
            warning(paste("Site", site_name,
                          "has a younger possible start date than 2500 BCE",
                          "with an isobase direction of",
                          unique(temp_curve$direction),
                          "and is returned as NA."))
          }
        }
      }


      # Normalise to sum to unity
      if (!(all(is.na(dategrid$probability))) & normalise) {
        dategrid$probability <- dategrid$probability / sum(dategrid$probability)
      }

      # Update list holding results
      if (!(all(is.na(dategrid$probability))) & !sparse) {

        hdr <- shoredate_hdr(dategrid$bce, dategrid$probability,
                             site_name, cal_reso, hdr_prob)

        date_isobases[[k]] <- list(
          site_name = site_name,
          site_elev = siteelev,
          date = dategrid,
          weighted_mean = hdr$weighted_mean,
          hdr_start = hdr$start,
          hdr_end = hdr$end,
          hdr_prob = hdr_prob,
          dispcurve = temp_curve[, names(temp_curve) %in%
                                   c("bce", "lowerelev", "upperelev")],
          dispcurve_direction = unique(temp_curve$direction),
          model_parameters = model_parameters,
          modeldat = modeldat,
          cal_reso = cal_reso)

      } else if (all(is.na(dategrid$probability)) & !sparse){
        date_isobases[[k]] <- list(
          site_name = site_name,
          site_elev = siteelev,
          date = dategrid,
          weighted_mean = NA,
          hdr_start = NA,
          hdr_end = NA,
          hdr_prob = hdr_prob,
          dispcurve = temp_curve[, names(temp_curve) %in%
                                   c("bce", "lowerelev", "upperelev")],
          dispcurve_direction = unique(temp_curve$direction),
          model_parameters = model_parameters,
          modeldat = modeldat,
          cal_reso = cal_reso)
      } else {
        date_isobases[[k]] <- dategrid
      }
    }

    if (sum_isobase_directions) {
      sum_isobases <- sum_shoredates(date_isobases, normalise = normalise)

      # Find the HDR for the sum
      hdr <- shoredate_hdr(sum_isobases$sum$bce, sum_isobases$sum$probability,
                           site_name, cal_reso, hdr_prob)

      date_isobases <- list(list(
        site_name = site_name,
        site_elev = siteelev,
        date = sum_isobases$sum,
        weighted_mean = hdr$weighted_mean,
        hdr_start = hdr$start,
        hdr_end = hdr$end,
        hdr_prob = hdr_prob,
        dispcurve = NA,
        dispcurve_direction = unlist(lapply(date_isobases,
                              function(x) unique(x["dispcurve_direction"]))),
        model_parameters = model_parameters,
        modeldat = modeldat,
        cal_reso = cal_reso
      ))

    }

    shorelinedates[[i]] <- date_isobases

  }

  class(shorelinedates) <- c("shoreline_date", class(shorelinedates))

  shorelinedates
}
